### ISSUE TITLE
scrollbar example 

### ISSUE THREAD

**Issue by mrtodd** [oooutlk/tcltk#10]
I am going through your tutorial and it appears the "elevator" is stuck while using the tclosure for the command.  I changed the code as follows and things are working for me:

```
    let s = root.add_ttk_scrollbar( "s" -orient("vertical") -command( ".l yview") )?
//    let s = root.add_ttk_scrollbar( "s" -orient("vertical") -command( tclosure!( tk, 
//                move |..| -> TkResult<(c_double, c_double)> {
//                    Ok( l.yview()? )
//                })))?
        .grid( -column(1) -row(0) -sticky("ns") )?;
```
In the original code the elevator moves when I use the mousewheel in the listbox, but when I try to move the elevator directly  with the mouse nothing happens (except a bunch of ".l yview" in the debug output).  The commented out code compiled.  Can you see the problem?

-- Todd

**Comment by oooutlk** [comment]
I have checked the source code and found the root cause of this issue. Will be fixed in several days.

I started this hobby project quite early, at the beginning of 2020 without publishing on github. It stalled for several times due to some personal reasons. When I continue to code, something might be forgotten. I think this situation caused your two issues #9 and #10. In issue #9 I forgot to add quite a lot keys in `TkKey`. In issue #10(this one) I forgot to utilize variadic arguments to complete the arguments of `yview()`.

**Comment by mrtodd** [comment]
My first look at Rust was last week, but I was introduced to Tcl/Tk in the early '90s and worked with many EDA tools where Tcl was included.  At this point I didn't feel I was ready to dig deep enough into the Rust code to attempt to track down this issue.
You might want to put "Take a look at section 8.4 of the tutorial" on your to-do list.  I think you'll agree that a bit of rework is needed here.  
I am trying to be helpful, but just let me know if you would like me to be more quiet.

-- Todd

    On Thursday, November 23, 2023 at 11:44:48 PM GMT+8, oooutlk ***@***.***> wrote:  
 
 


I have checked the source code and found the root cause of this issue. Will be fixed in several days.

I started this hobby project quite early, at the beginning of 2020 without publishing on github. It stalled for several times due to some personal reasons. When I continue to code, something might be forgotten. I think this situation caused your two issues #9 and #10. In issue #9 I forgot to add quite a lot keys in TkKey. In issue #10(this one) I forgot to utilize variadic arguments to complete the arguments of yview().

**Comment by oooutlk** [comment]
Thank your efforts to help improving this project! I'm more familiar with Rust rather than Tcl/Tk, and experienced Tcl/Tk programmers are welcome to contribute, find bugs or make sugguestion,  to the Tcl/Tk part( or the Rust binding part ).

The [tutorial](https://oooutlk.github.io/tk/) of this project is "translated to Rust" from [tkdocs](https://tkdocs.com/tutorial/). Did you mean [section 8.4](https://oooutlk.github.io/tk/threads_or_processes.html) has some issue? Its upstream page is [here](https://tkdocs.com/tutorial/eventloop.html#threads). Please tell me if it needs some improvement.

I'd like to mention something worth noting:

1. The users can run Tcl scripts and write Tcl extentions in Rust. Some glue code generated by Rust proc_macros help to fill the gap between Tcl proc and Rust function/closure.
2. Chained subtraction of Rust expressions to simulate Tk options.
3. Automated hierarchical paths make the code more compact.
4. Respect to corresponding Tcl/Tk's API's naming, to make it easier for Tcl/Tk programs to use this project.
5. Something more than Tcl/Tk. For instance, widget tree w/o geometry in one expression, option valid-check at compile time, and the ext mod.

And something that needs review:

1. The crate tk is more a "Tk Commands" binding library rather than a TkLib binding library. Do we need to write more C API bindings of TkLib recently?
2. The crate tcl provides only necessary APIs to support Tk. Do we need to provide other APIs recently?
3. Any bug or misunderstanding in the low level part implementation of Tcl bindings?

**Comment by mrtodd** [comment]
The tutorial section I called out has a problem with parts of it being repeated.  I noticed this happened again in another section as I continued to read.  I don't know if you are in the US, but reading section 8.4 was similar to listening to one of VP Harris' speeches.  Do you have a repository for the tutorial?  I could go through it and submit the changes back to you.
I need to re-read and think about the rest of your message below -- perhaps, after my daughter goes to bed and the house gets a bit more quiet.

-- Todd

    On Friday, November 24, 2023 at 03:08:53 PM GMT+8, oooutlk ***@***.***> wrote:  
 
 


Thank your efforts to help improving this project! I'm more familiar with Rust rather than Tcl/Tk, and experienced Tcl/Tk programmers are welcome to contribute, find bugs or make sugguestion, to the Tcl/Tk part( or the Rust binding part ).

The tutorial of this project is "translated to Rust" from tkdocs. Did you mean section 8.4 has some issue? Its upstream page is here. Please tell me if it needs some improvement.

I'd like to mention something worth noting:
   
   - The users can run Tcl scripts and write Tcl extentions in Rust. Some glue code generated by Rust proc_macros help to fill the gap between Tcl proc and Rust function/closure.
   - Chained subtraction of Rust expressions to simulate Tk options.
   - Automated hierarchical paths make the code more compact.
   - Respect to corresponding Tcl/Tk's API's naming, to make it easier for Tcl/Tk programs to use this project.
   - Something more than Tcl/Tk. For instance, widget tree w/o geometry in one expression, option valid-check at compile time, and the ext mod.

And something that needs review:
   
   - The crate tk is more a "Tk Commands" binding library rather than a TkLib binding library. Do we need to write more C API bindings of TkLib recently?
   - The crate tcl provides only necessary APIs to support Tk. Do we need to provide other APIs recently?
   - Any bug or misunderstanding in the low level part implementation of Tcl bindings?

**Comment by oooutlk** [comment]
The markdown source files are part of this repository, under tk/book/src directory. Section 8.4 is [here)[https://github.com/oooutlk/tcltk/blob/main/tk/book/src/threads_or_processes.md]

I have found these repeated statements. Sorry for speak in the Harris way :)

**Comment by oooutlk** [comment]
I published tcl_derive 0.1.3 and tk 0.1.8. This issue has been fixed.

Notice that the newly added `yview()` is replaced by `yview_()` in the following code.

Frankly speaking, it's a bit verbose than '.l view'.

Just to demonstrate another way of doing it, using closures, macros and predefined functions rather than command prefix strings.

```
    let s = root.add_ttk_scrollbar( "s"
            -orient("vertical")
            -command( tclosure!( tk,
                move |..| -> TkResult<()> { Ok( l.yview_( tcl_va_args!() )? )})))?
        .grid( -column(1) -row(0) -sticky("ns") )?;
```
